lightningd/invoice.c

```diff
@@ -1441,3 +1475,162 @@ static const struct json_command decodepay_command = {
 	"Decode {bolt11}, using {description} if necessary"
 };
 AUTODATA(json_command, &decodepay_command);
+
+/* If we fail because it exists, we also return the clashing invoice */
+static struct command_result *fail_exists(struct command *cmd,
+					  const struct json_escape *label)
+{
+	struct json_stream *data;
+	struct invoice invoice;
+	struct wallet *wallet = cmd->ld->wallet;
+
+	data = json_stream_fail(cmd, INVOICE_LABEL_ALREADY_EXISTS,
+				"Duplicate label");
+	if (!wallet_invoice_find_by_label(wallet, &invoice, label))
+		fatal("Duplicate invoice %s not found any more?",
+		      label->s);
+
+	json_add_invoice(data, wallet_invoice_details(cmd, wallet, invoice));
+	json_object_end(data);
+
+	return command_failed(cmd, data);
+}
+
+static struct command_result *json_createinvoice(struct command *cmd,
+						 const char *buffer,
+						 const jsmntok_t *obj UNNEEDED,
+						 const jsmntok_t *params)
+{
+	const char *invstring;
+	struct json_escape *label;
+	struct preimage *preimage;
+	struct invoice invoice;
+	struct sha256 payment_hash;
+	struct json_stream *response;
+	struct bolt11 *b11;
+	struct sha256 hash;
+	u5 *sig;
+	bool have_n;
+	char *fail;
+
+	if (!param(cmd, buffer, params,
+		   p_req("invstring", param_string, &invstring),
+		   p_req("preimage", param_preimage, &preimage),
```

Comment:
Can we make the `preimage` optional? That way we can build arbitrary invoice handling structures using `htlc_accepted` hooks, that do not require internal resolution, but still require a valid signature from the node?

I can think of two scenarios that this would be useful right away:
 - Create fake MPP-probe invoices that we do not have preimages for (and want to prove that's the case by using a derivation that doesn't involve a preimage such as "xor two hashes with each other")
 - Have an external invoice management system, in which the node is untrusted and requires an incoming payment be acknowledged by the external system (accounting, compliance, ....)

Reply:
That's a bigger change; we'd have to allow the preimage entry to be null in the db.  However, I'll reorder the args so we can make it optional in future more easily?